### ISSUE TITLE
HashSet name conflict with Generic.HashSet (2.0-branch)

### ISSUE THREAD

**Issue by StefanBertels** [louthy/language-ext#193]
Error	CS0104	'HashSet<>' is an ambiguous reference between 'System.Collections.Generic.HashSet<T>' and 'LanguageExt.HashSet<A>'

What's the suggested solution to avoid this naming conflict?
(Why is languageext-2.0-branch HashSet not named HSet?)

**Comment by louthy** [comment]
The name `HSet` and `HMap` was really annoying me, it's ugly and compromised by a worse implementation (IMHO).  I didn't like the compromise that was inflicted by Visual Studio's default includes (when this library very often removes the need for the collections in `System.Collections.Generic`).

The simplest solution I've found for dealing with this is:

* Remove `using System.Collections.Generic`
* Replace with `using G = System.Collections.Generic`

Then use the `G` prefix for types required from the `System.Collections.Generic` namespace:
```c#
        G.HashSet<int> a     = new G.HashSet<int>();
        G.IEnumerable<int> b = Enumerable.Range(0, 100);
        HashSet<int> c       = HashSet(1, 2, 3);
        Lst<int> d           = List(1, 2, 3);
```

**Comment by StefanBertels** [comment]
Hi Paul,

ok. I understand your motivation and agree that LanguageExt supplies the better Collection types. I'm just not sure whether this namespace conflict will help people converting. You now have to refactor your classes if you need/want co-existence. Just a note -- it's not a problem for me.

I wondered because the list type is names `Lst `and only the Prelude has a `List` factory (which does not conflict with `new List()` of Collections.Generic) -- whereas the hash set now uses `HashSet `as type name.
Do you plan to change `Lst `to `List`, too?

Another thought on this: Methods like `Add` or `Contains` tend to have the same signature for `G.HashSet` and your `HashSet `-- but different behaviour. If you change an existing code-base by just changing the using directive, a line like:

`myHashSet.Add(...);`

does not work as (originally) intended with your immutable `HashSet `(return value unused). Different type names would avoid "auto-converting" methods that use HashSet in signature (without fully qualified namespace). 

If I think about this more, it might even be safer to not solve conflict by `using G = ...` but to fully qualify all usages until Collections.Generic is no longer used in class.

Anyway: 
I agree that type names like HSet, Lst etc. are ugly. So this is just a comment -- I don't have a better idea and can work with either way.
Thank you again for your great lib.

**Comment by StefanBertels** [comment]
related to #118

**Comment by louthy** [comment]
> Do you plan to change `Lst` to `List`, too?

No.  I took the plunge with `HSet` and `HMap` because they were both very new types in the library, and were likely to be used much less (there is very little benefit to using them other than supporting unorderable keys).

Although I'd prefer to use `List<A>` rather than `Lst<A>`, I did notice some conflicts in early versions (I can't remember exactly what the issue was now though!).  So I went for `Lst<A>`.  I also did the same with `Stck<A>` instead of `Stack<A>`, and `Que<A>` instead of `Queue<A>`.  _(`Que<A>` sets off my spelling-OCD like you wouldn't believe!)_

The one thing with all of those contracted names is this: it's still very clear what they are.  The problem with `HMap` and `HSet` is that it's unclear what the `H` means.  Obviously it doesn't take long to learn, but it's better if you can just look at the code and know what's going on.

And with `HashMap` available, I figured it would be worth the pain to rename both `HMap` to `HashMap` and `HSet` to `HashSet`.

I am always open to suggestions.  I am at the `alpha` stage with `v2`, so this is definitely the time to get the ideas to me :)

The only other ones I considered was `HshMap` and `HshSet`.  But they _feel_ really awkward.

**Comment by StefanBertels** [comment]
I don't see a general difference between List, Stack, Queue and HashSet here. 
I saw you added it later (see #111) but I don't see a difference in principal.

Current naming feels a bit arbitrary. I don't blame you. Naming is hard -- and this special problem is very hard. I suggest to strategically decide which way to go. Currently a (breaking) change in v2 won't hurt that much.

I really hope you will get more users, so a breaking change will hurt much more later.
There is probably more work to be done when switching existing code completely away from Collections.Generic than with something like changing Lst to List.

If you feel bad about using List maybe this is a reason to avoid naming collision (follow the ugly path).
Or do you think HashSet is much less important and conflicts there are (much) more acceptable?

The ugly path would introduce the next ugly question: Should it be `var list = Lst(1,2,3) // Prelude`?
There seems to be no technical problem, mixing Lst struct with List method works ok. But feels a bit strange (until one sees the reason).

I think there is a high value with collision-free usage. Will be easier for users to introduce your lib step by step in practise.

Hmm. 
**I**'m ready for change and use the nice names everywhere -- removing System.Collections.Generic.

Side node "Array": 

There is some other collision with System.Array. I just found an old method of mine with `Array.ConvertAll` and `Array.FindIndex` in my code. The compiler complains that Array is a method as found in Prelude ("Cannot choose method from method group"). The message gave no hint that there is a System.Array class...  Maybe those static methods won't be used that often -- I personally would writer "better" code today.

**Comment by louthy** [comment]
>  don't see a general difference between List, Stack, Queue and HashSet here.

I don't remember exactly what the problem was tbh, I just remember there being an issue that made it more awkward than I'd have liked. But you're right that the issues should be the same for all types.

> Or do you think HashSet is much less important and conflicts there are (much) more acceptable?

Not all breaking changes are equal.  Yes, version 2 is my opportunity to fix some legacy issues as well as bring in a raft of cool stuff (and I think people understand that major version changes might have breaking changes).  But I know that certain types will be used more than others.  `Option`, `Lst`, and `Map` probably get the heaviest 'hit' of any code in the library, so changing those is very problematic.  `HashSet` is just likely to be used less, and therefore I am less worried about naming clashes with the BCL.  But C# does have a way of dealing with naming clashes, and that's to namespace or alias.  I realise it's not pretty, but on this one type I will compromise because I think `HashSet` is the best name.

> The compiler complains that Array is a method as found in Prelude ("Cannot choose method from method group"). 

I'd like to lobby the Roslyn team to sort this out - I think it's a bug.  It's impossible to _dot into_ a method-group, so the compiler should know that `Array.ConvertAll` in this instance refers to the type or variable called `Array`.  It's also impossible to pass arguments to a type:  `Array(1,2,3)` is obviously calling a method.  I think it may have been missed when the Roslyn team implemented `using static` for C#6.

This is an issue in lang-ext now with the `Map` function in `Prelude` (for example).  If you do `using LanguageExt.Predude` then you can construct a map with `Map((1,2),(3,4))`.  But the `Map` static type can't be accessed: `Map.create<int, int>()`.

Unfortunately there are a few compromises, but I think overall the balance is about right at the moment.
